docs(openid-connect): docs improvements (#5096)
Kong/docs.konghq.com: app/_hub/kong-inc/openid-connect/_index.md

publisher: Kong Inc.
desc: Integrate Kong with a third-party OpenID Connect provider
description: |

>   OpenID Connect ([1.0][connect]) plugin allows the integration with a third party

Reviewer: ```suggestion
  OpenID Connect ([1.0][connect]) plugin allows for integration with a third party
```

  ## Important Configuration Parameters

>   This plugin includes many configuration parameters that allow a finely grained customization.
>   The following steps will help you getting started setting up the plugin:

Reviewer: ```suggestion
  This plugin includes many configuration parameters that allow finely grained customization.
  The following steps will help you get started setting up the plugin:
```

  1. Configure: `config.issuer`.

     This parameter tells the plugin where to find discovery information, and it is
     the only required parameter. You should set the value `realm` or `iss` on this
     parameter if you don't have a discovery endpoint.

     {:.note}
      > **Note**: This does not have
     to match the URL of the `iss` claim in the access tokens being validated. To set
     URLs supported in the `iss` claim, use `config.issuers_allowed`.
  2. Decide what authentication grants to use with this plugin and configure

>      the field `config.auth_methods` accordingly.

Reviewer: ```suggestion
     the `config.auth_methods` field accordingly.
```

>      The parameter should only contain the grants that you want to use, in order to
>      restrict the scope of any potential attack.

Reviewer: ```suggestion
     In order to restrict the scope of potential attacks, the parameter should only contain the grants that you want to use. 
```

     the audience with `config.audience_required` to contain only your `config.client_id`.
     You may also need to adjust `config.audience_claim` in case your identity provider
     uses a non-standard claim (other than `aud` as specified in JWT standard). This is

>      important because some identity providers, such as Google, share the public keys

Reviewer: ```suggestion
     important because some identity providers, such as Google, share public keys
```

     with different clients.

>   5. If you are using Kong in DB-less mode with the declarative configuration and 
>      session cookie authentication, you should set `config.session_secret` as well.
>      Leaving this parameter unset would result in each of the Nginx workers across your
>      nodes encrypting and signing the cookies with their own secrets.

Reviewer: ```suggestion
  5. If you are using Kong in DB-less mode with a declarative configuration and 
     session cookie authentication, you should set `config.session_secret`.
     Leaving this parameter unset will result in every Nginx worker across your
     nodes encrypting and signing the cookies with their own secrets.
```

      default: null
      datatype: string
      description:

>         An optional string (consumer UUID or username) value to use as an “anonymous” consumer if authentication fails. If empty (default null), requests that fail authentication will return a `4xx` status code. Note that this value must refer to the consumer `id` or `username` attribute, and **not** its `custom_id`.

Reviewer: ```suggestion
        An optional string (consumer UUID or username) value that functions as an “anonymous” consumer if authentication fails. If empty (default null), requests that fail authentication will return a `4xx` HTTP status code. This value must refer to the consumer `id` or `username` attribute, and **not** its `custom_id`.
```

        - `PS512`: RSASSA-PSS using SHA-512 and MGF1 with SHA-512
        - `EdDSA`: EdDSA with Ed25519
    - group: JWT Access Token Authentication

>       description: Parameters for setting where to look for the bearer token and whether to introspect them.

Reviewer: ```suggestion
      description: Parameters for specifying the location of the bearer token, and introspection options.
```

      datatype: boolean
      description: Specifies whether to introspect the JWT access tokens (can be used to check for revocations).
    - group: Client Credentials Grant

>       description: Parameters for where to look for the client credentials.

Reviewer: ```suggestion
      description: Parameters specifying the location of client credentials.
```

      default:
        - sub
      datatype: array of string elements

>       description: 'The claim used to derive a virtual credential (e.g. to be consumed by the rate-limiting plugin), in case the Consumer mapping is not used.'

Reviewer: ```suggestion
      description: 'The claim used to derive virtual credentials (e.g. to be consumed by the rate-limiting plugin), in case the consumer mapping is not used.'
```

but the steps will be similar in other standard identity providers. If you encounter
difficulties during this phase, please refer to the [Keycloak documentation](https://www.keycloak.org/documentation).

1. Create a confidential client `kong` with `private_key_jwt` authentication and configure
   Keycloak to download the public keys from [the OpenID Connect Plugin JWKS endpoint][json-web-key-set]:
   <br><br>
   <img src="/assets/images/docs/openid-connect/keycloak-client-kong-settings.png">
   <br>
   <img src="/assets/images/docs/openid-connect/keycloak-client-kong-auth.png">
   <br>
2. Create another confidential client `service` with `client_secret_basic` authentication.
   For this client, Keycloak will auto-generate a secret similar to the following: `cf4c655a-0622-4ce6-a0de-d3353ef0b714`.
   Enable the client credentials grant for the client:
   <br><br>
   <img src="/assets/images/docs/openid-connect/keycloak-client-service-settings.png">
   <br>
   <img src="/assets/images/docs/openid-connect/keycloak-client-service-auth.png">
   <br>

> 3. Create a verified user with name: `john` and the non-temporary password: `doe` that can be used with the password grant:

Reviewer: ```suggestion
3. Create a verified user with the name: `john` and the non-temporary password: `doe` that can be used with the password grant:
```

Using Admin API is convenient when testing the plugin, but similar configs can
be done in declarative format as well.

> When this plugin is configured with multiple grants / flows there is a hard-coded search

Reviewer: ```suggestion
When this plugin is configured with multiple grants/flows there is a hard-coded search
```

8. [Client Credentials Grant](#client-credentials-grant)
9. [Authorization Code Flow](#authorization-code-flow)

> Once some credentials are found, the plugin will stop searching further. Multiple grants may

Reviewer: ```suggestion
Once credentials are found, the plugin will stop searching further. Multiple grants may
```

> share the same credentials, i.e. both the password and client credentials grants can use 

Reviewer: ```suggestion
share the same credentials. For example, both the password and client credentials grants can use 
```

basic access authentication through the `Authorization` header.

{:.warning}
> The choices made in the examples below are solely aimed at simplicity.

> Because httpbin.org is used as an upstream service, it is highly recommend that you do

Reviewer: ```suggestion
Because `httpbin.org` is used as an upstream service, it is highly recommended that you do
```

### Password Grant

> Password grant is a legacy authentication grant. This is a less secure way of
> authenticating end users than the authorization code flow, because, for example,
> the passwords are shared with third parties. The grant is rather simple:

Reviewer: ```suggestion
Password grant is a legacy authentication grant. This is a less secure way of
authenticating end users than the authorization code flow, because, for example,
the passwords are shared with third parties. The image below illustrates the grant:
```

### Client Credentials Grant

> Client credentials grant is very similar to [the password grant](#password-grant).

Reviewer: ```suggestion
The client credentials grant is very similar to [the password grant](#password-grant).
```

## Debugging

> The OpenID Connect plugin is complex, and it integrates with a third party
> identity provider: this can present some challenges while debugging it. If you have

Reviewer: ```suggestion
The OpenID Connect plugin is complex, integrating with third-party identity providers can present challenges. If you have
```
